pkg/engine/task/task_kudo_operator.go

```diff
@@ -11,9 +30,165 @@ type KudoOperatorTask struct {
+	vals := renderer.
+		NewVariableMap().
+		WithInstance(meta.OperatorName, meta.InstanceName, meta.InstanceNamespace, meta.AppVersion, meta.OperatorVersion).
+		WithParameters(parameters)
+
+	engine := renderer.New()
+
+	return engine.Render(pf, pft, vals)
+}
+
+func instanceResource(instanceName, operatorName, operatorVersionName, namespace string, parameters map[string]string, owner metav1.Object, scheme *runtime.Scheme) (*v1beta1.Instance, error) {
+	instance := &v1beta1.Instance{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "Instance",
+			APIVersion: packages.APIVersion,
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      instanceName,
+			Namespace: namespace,
+			Labels:    map[string]string{kudo.OperatorLabel: operatorName},
+		},
+		Spec: v1beta1.InstanceSpec{
+			OperatorVersion: corev1.ObjectReference{
+				Name: operatorVersionName,
+			},
+			Parameters: parameters,
+		},
+		Status: v1beta1.InstanceStatus{},
+	}
+	if err := controllerutil.SetControllerReference(owner, instance, scheme); err != nil {
+		return nil, fmt.Errorf("failed to set resource ownership for the new instance: %v", err)
+	}
+
+	return instance, nil
+}
+
+// applyInstance creates the passed instance if it doesn't exist or patches the existing one. Patch will override
+// current spec.parameters and Spec.operatorVersion the same way, kudoctl does it. If the was no error, then the passed
+// instance object is updated with the content returned by the server
+func applyInstance(new *v1beta1.Instance, ns string, c client.Client) error {
+	old := &v1beta1.Instance{}
+	err := c.Get(context.TODO(), types.NamespacedName{Name: new.Name, Namespace: ns}, old)
+
+	switch {
+	// 1. if instance doesn't exist, create it
+	case apierrors.IsNotFound(err):
+		log.Printf("Instance %s/%s doesn't exist. Creating it", new.Namespace, new.Name)
+		return createInstance(new, c)
+	// 2. if the instance exists (there was no error), try to patch it
+	case err == nil:
+		log.Printf("Instance %s/%s already exist. Patching it", new.Namespace, new.Name)
+		return patchInstance(new, c)
+	// 3. any other error is treated as transient
+	default:
+		return fmt.Errorf("failed to check if instance %s/%s already exists: %v", new.Namespace, new.Name, err)
+	}
+}
+
+func createInstance(i *v1beta1.Instance, c client.Client) error {
+	gvk := i.GroupVersionKind()
+	err := c.Create(context.TODO(), i)
+
+	// reset the GVK since it is removed by the c.Create call
+	// https://github.com/kubernetes/kubernetes/issues/80609
+	i.SetGroupVersionKind(gvk)
+
+	return err
+}
+
+func patchInstance(i *v1beta1.Instance, c client.Client) error {
+	patch, err := json.Marshal(struct {
+		Spec *v1beta1.InstanceSpec `json:"spec"`
+	}{
+		Spec: &i.Spec,
+	})
+
+	if err != nil {
+		return fmt.Errorf("failed to serialize instance %s/%s patch: %v", i.Namespace, i.Name, err)
+	}
+
+	return c.Patch(context.TODO(), i, client.RawPatch(types.MergePatchType, patch))
```

Comment:
I'm pretty sure this is not enough to correctly patch the instance, I want to see at least a test that removes a parameter from the updated instance. This could for example happen if the `paramsFile` for the dependent operator has something like:

```
{{ if eq ".Params.INCLUDE_STUFF }}
FANCY_PARAM_THAT_CAN_BE_THERE_OR NOT: "cool value"
{{ end }}
```

Reply:
Can we utilize the `applyResources` from the `task_apply.go`?

Reply:
You're right. I've adopted this logic from [kudo.go::UpdateInstance](https://github.com/kudobuilder/kudo/blob/master/pkg/kudoctl/util/kudo/kudo.go#L209). But switching to the `StrategicMergePatchType` will require storing the previous state but instance controller is already doing that... I've created an issue #1544

Reply:
The `applyResources` sets all the labels and annotations but more importantly, it sets the `kudo.dev/last-applied-configuration`. However, an Instance already has it set by the `instance_controller.go` called `kudo.dev/last-applied-instance-state`, which is something I wanted to removed altogether. What do you think about setting `patchStrategy:"merge"` for the `spec.parameters` like:
```
type InstanceSpec struct {
	...
	Parameters map[string]string `json:"parameters,omitempty" patchStrategy:"merge"`
}
```
At least in the `kudo_operator_task.go` we generate a full map of parameters each time so overriding everything should work, right?